### PR TITLE
logger

### PR DESCRIPTION
**Pull request by JainishaParmar**
added winston logger

### CODE REVIEW

**Review by jaimin45 — APPROVED** [review]

**Review by jaimin45 — APPROVED** [review]

File: app.js [hunk 1]
```diff
@@ -13,9 +13,13 @@ const dirname = path.resolve();
 
 const app = express();
 const swaggerDoc = swaggerYaml.load('./swagger.yaml');
+const tutorialWinston = require("./config/tutorial.winston");
 
-// eslint-disable-next-line no-console
-promise.then(() => console.log("Connection successful!"));
+if (promise) {
```
Comment (kapilbarad) [inline app.js]: This is not how a promise works. Please read this docs:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

CC: @ashishkava @jaimin45

File: app.js [hunk 2]
```diff
@@ -13,9 +13,13 @@ const dirname = path.resolve();
 
 const app = express();
 const swaggerDoc = swaggerYaml.load('./swagger.yaml');
+const tutorialWinston = require("./config/tutorial.winston");
```
Comment (kapilbarad) [inline app.js]: We don't need to define loggers based on modules.

There will be only one logger `./config/logger`, Which should export a logger instance by default.
So we can use it like this:

```
logger.info('Something....');
```

CC: @jaimin45 @ashishkava

File: config/tutorial.winston.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      format: format.combine(format.timestamp(), format.json()),
+    }),
+    new transports.Console({
+      level: "error",
+      format: format.combine(format.timestamp(), format.json()),
+    }),
+  ],
+});
+
+module.exports = { tutorialLogger };
```
Comment (kapilbarad) [inline config/tutorial.winston.js]: We should export logger as default. Also name of the variable should be `logger`.

File: controllers/tutorial.controller.js [hunk 4]
```diff
@@ -71,14 +77,18 @@ const deleteTutorial = async (req, res) => {
 
     if (tutorial === null) {
       res.status(404).send({ message: "Tutorial not found" });
+      logger.tutorialLogger.log('error', 'Tutorial not found');
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We only need to log 500 (any 5xx errors) errors. No need to log 400, 404, or any other 4xx errors.

CC: @jaimin45 @ashishkava

File: package.json [hunk 5]
```diff
@@ -21,6 +21,7 @@
     "prepare": "husky install"
   },
   "dependencies": {
+    "app-root-path": "^3.0.0",
```
Comment (kapilbarad) [inline package.json]: What is the need of this dependency - `app-root-path`?

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by jaimin45 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: controllers/tutorial.controller.js [hunk 6]
```diff
@@ -11,31 +12,29 @@ const getTutorials = async (req, res) => {
...
     }
+    const tutorial = new Tutorial(value);
+    await tutorial.save();
+    res.status(201).send();
   } catch (error) {
     if (error instanceof CastError) {
       res.status(400).send({ message: "Invalid Tutorial details" });
     } else {
       res.status(500).send({ message: "Internal Server Error" });
+      logger.error('Internal Server Error');
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We also need to log actual error, otherwise how would we know that what was the root cause of this 500?

Comment (kapilbarad) [reply controllers/tutorial.controller.js]: You can configure this method to print errors:

```
const winston = require('winston');
const { format } = require('logform');

const logger = winston.createLogger({
  format: format.combine(
    format.errors({ stack: true }),
    format.metadata(),
    format.json(),
  ),
  transports: [ new winston.transports.Console() ],
});
```
Now you can log errors like this:
```
logger.error('Could's create tutorial: ', error);
```


For above code to work, You need to install `logform` dependency.

CC: @jaimin45 @JainishaParmar @ashishkava

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by jaimin45 — APPROVED** [review]

File: controllers/tutorial.controller.js [hunk 7]
```diff
@@ -10,32 +11,30 @@ const getTutorials = async (req, res) => {
     });
     res.status(200).send(tutorials);
   } catch (error) {
+    logger.error('Internal server error', error);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: Error name should be something we can easily track: `('Couldn't fetch tutorials: ', error)`

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Comment by JainishaParmar** [comment]
feat: winston logger

**Comment by JainishaParmar** [comment]
feat: winston logger

**Pull request merged**